gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,431 @@
+    cleanup_count: usize,
+}
+
+impl DuplicateShredHandlerTrait for DuplicateShredHandler {
+    // Here we are sending data one by one rather than in a batch because in the future
+    // we may send different type of CrdsData to different senders.
+    fn handle(&mut self, shred_data: DuplicateShred) {
+        if let Err(error) = self.handle_shred_data(shred_data) {
+            error!("handle packet: {:?}", error)
+        }
+        if self.cleanup_count.saturating_sub(1) == 0 {
+            self.cleanup_old_slots();
+            self.cleanup_count = CLEANUP_EVERY_N_LOOPS;
+        }
+    }
+}
+
+impl DuplicateShredHandler {
+    pub fn new(
+        blockstore: Arc<Blockstore>,
+        leader_schedule_cache: Arc<LeaderScheduleCache>,
+    ) -> Self {
+        Self {
+            chunk_map: HashMap::new(),
+            validator_pending_proof_map: HashMap::new(),
+            last_root: 0,
+            blockstore,
+            leader_schedule_cache,
+            cleanup_count: CLEANUP_EVERY_N_LOOPS,
+        }
+    }
+
+    fn handle_shred_data(&mut self, data: DuplicateShred) -> Result<(), Error> {
+        if self.should_insert_chunk(&data) {
+            let slot = data.slot;
+            match self.insert_chunk(data) {
+                Err(error) => return Err(error),
+                Ok(Some(chunks)) => {
+                    self.verify_and_apply_proof(slot, chunks)?;
+                    // We stored the duplicate proof in this slot, no need to accept any future proof.
+                    self.mark_slot_proof_received(slot);
+                }
+                _ => (),
+            }
+        }
+        Ok(())
+    }
+
+    fn should_insert_chunk(&self, data: &DuplicateShred) -> bool {
+        let slot = data.slot;
+        // Do not insert if this slot is rooted or has a proof already.
+        if slot <= self.blockstore.last_root() || self.blockstore.has_duplicate_shreds_in_slot(slot)
+        {
+            return false;
+        }
+        // Discard all proofs with abnormal num_chunks.
+        if data.num_chunks == 0 || data.num_chunks > MAX_NUM_CHUNKS {
+            return false;
+        }
+        // Only allow limited unfinished proofs per pubkey to reject attackers.
+        if let Some(current_slots_set) = self.validator_pending_proof_map.get(&data.from) {
+            if !current_slots_set.contains(&slot)
+                && current_slots_set.len() >= ALLOWED_SLOTS_PER_PUBKEY
+            {
+                return false;
+            }
+        }
+        // Also skip frozen slots or slots with a newer proof than me.
+        match self.chunk_map.get(&slot) {
+            Some(SlotStatus::Frozen) => {
+                return false;
+            }
+            Some(SlotStatus::UnfinishedProof(slot_map)) => {
+                if let Some(proof_chunkmap) = slot_map.get(&data.from) {
+                    if proof_chunkmap.wallclock > data.wallclock {
+                        return false;
+                    }
```

Comment:
This is to prevent spamming. From from_duplicate_slot_proof(), you can send as many duplicate proofs as you want, but each proof will have a unique wallclock which is carried by all chunks. We only accept one duplicate proof to limit the size of the map.

Reply:
so we always pick the one with the higher wallclock? seems it would be better to pick the lower wallclock to converge quicker